src/firmware/app/primitives/shoot_alignment.c

```diff
@@ -0,0 +1,108 @@
+#include "firmware/app/primitives/shoot_alignment.h"
+
+#include "firmware/app/control/bangbang.h"
+#include "firmware/app/control/control.h"
+#include "firmware/app/control/physbot.h"
+#include "firmware/app/primitives/primitive.h"
+#include "firmware/shared/physics.h"
+#include "firmware/shared/util.h"
+
+#define TIME_HORIZON 0.05f  // in seconds
+
+typedef struct ShootAlignmentState
+{
+    float destination[3], major_vec[2], minor_vec[2], total_rot;
+} ShootAlignmentState_t;
+DEFINE_PRIMITIVE_STATE_CREATE_AND_DESTROY_FUNCTIONS(ShootAlignmentState_t)
+
+/**
+ * Scales the major acceleration by the distance from the major axis and the
+ * amount required left to rotate.
+ *
+ * @pre Total rotation and the distance vector should
+ * not be zero so as to avoid divide by zero errors.
+ *
+ * @param pb [in/out] The PhysBot data container that contains information about the
+ * major and minor axis.
+ */
+void scale(PhysBot *pb)
+{
+    float maj_disp        = fabsf(pb->maj.disp) - ROBOT_RADIUS;
+    float distance_vector = sqrtf(powf(maj_disp, 2) + powf(pb->min.disp, 2));
+    if (distance_vector != 0)
+    {
+        float abs_factor        = fabsf(pb->min.disp) / distance_vector;
+        float minor_axis_factor = 1 - abs_factor;
+        pb->maj.accel *= minor_axis_factor;
+    }
+}
+
+/**
+ * Determines the rotation acceleration after setup_bot has been used and
+ * plan_move has been done along the minor axis. The minor time from bangbang
+ * is used to determine the rotation time, and thus the rotation velocity and
+ * acceleration. The rotational acceleration is clamped under the MAX_T_A.
+ *
+ * @param pb [in/out] The PhysBot data container that should have minor axis time and
+ * will store the rotational information
+ * @param avel The rotational velocity of the bot
+ */
+void plan_shoot_rotation(PhysBot *pb, float avel)
+{
+    pb->rot.time = (pb->min.time > TIME_HORIZON) ? pb->min.time : TIME_HORIZON;
+    // 1.6f is a magic constant
```

Comment:
I assume this is out of scope since it's copied over but I'm not sure how I feel about "magic constants".